hpl-2.2/man/man3/HPL_pdlange.3
.TH HPL_pdlange 3 "February 24, 2016" "HPL 2.2" "HPL Library Functions"
.SH NAME
HPL_pdlange \- Compute ||A||.
.SH SYNOPSIS
\fB\&#include "hpl.h"\fR
 
\fB\&double\fR
\fB\&HPL_pdlange(\fR
\fB\&const HPL_T_grid *\fR
\fI\&GRID\fR,
\fB\&const HPL_T_NORM\fR
\fI\&NORM\fR,
\fB\&const int\fR
\fI\&M\fR,
\fB\&const int\fR
\fI\&N\fR,
\fB\&const int\fR
\fI\&NB\fR,
\fB\&const double *\fR
\fI\&A\fR,
\fB\&const int\fR
\fI\&LDA\fR
\fB\&);\fR
.SH DESCRIPTION
\fB\&HPL_pdlange\fR
returns  the value of the one norm,  or the infinity norm,
or the element of largest absolute value of a distributed matrix A:  
 
 
   max(abs(A(i,j))) when NORM = HPL_NORM_A,                          
   norm1(A),        when NORM = HPL_NORM_1,                          
   normI(A),        when NORM = HPL_NORM_I,                          
 
where norm1 denotes the one norm of a matrix (maximum column sum) and
normI denotes  the infinity norm of a matrix (maximum row sum).  Note
that max(abs(A(i,j))) is not a matrix norm.
.SH ARGUMENTS
.TP 8
GRID    (local input)           const HPL_T_grid *
On entry,  GRID  points  to the data structure containing the
process grid information.
.TP 8
NORM    (global input)          const HPL_T_NORM
On entry,  NORM  specifies  the  value to be returned by this
function as described above.
.TP 8
M       (global input)          const int
On entry,  M  specifies  the number  of rows of the matrix A.
M must be at least zero.
.TP 8
N       (global input)          const int
On entry,  N specifies the number of columns of the matrix A.
N must be at least zero.
.TP 8
NB      (global input)          const int
On entry,  NB specifies the blocking factor used to partition
and distribute the matrix. NB must be larger than one.
.TP 8
A       (local input)           const double *
On entry,  A  points to an array of dimension  (LDA,LocQ(N)),
that contains the local pieces of the distributed matrix A.
.TP 8
LDA     (local input)           const int
On entry, LDA specifies the leading dimension of the array A.
LDA must be at least max(1,LocP(M)).
.SH SEE ALSO
.BR HPL_pdlaprnt \ (3),
.BR HPL_fprintf \ (3).


hpl-2.2/HISTORY
==============================================================
 High Performance Computing Linpack Benchmark (HPL)
 HPL - 2.2 - February 24, 2016
==============================================================

 History

 - 09/09/00 Public release of Version 1.0

 - 09/27/00 A couple of mistakes in the  VSIPL  port have been
 corrected.  The tar file as well as the web site were updated
 on September 27th, 2000.  Note  that  these problems were not
 affecting the BLAS version of the software in any way.

 - 01/01/04 Version 1.0a
 The  MPI  process grid  numbering  scheme  is now an run-time
 option.
 The inlined assembly  timer  routine that caused the compila-
 tion to fail when using  gcc  version 3.3  and above has been
 removed from the package.
 Various building problems on the T3E have been fixed;  Thanks
 to Edward Anderson.

 - 15/12/04 Version 1.0b
 Weakness of the pseudo-random matrix generator found for pro-
 blem sizes being power of twos and larger  than 2^15;  Thanks
 to Gregory Bauer. This problem has not been fixed. It is thus
 currently recommended to  HPL  users willing to test matrices
 of size larger than 2^15 to not use power twos.

 When the matrix size is such that one needs  > 16 GB  per MPI
 rank,  the  intermediate  calculation  (mat.ld+1) * mat.nq in
 HPL_pdtest.c  ends up  overflowing  because  it is done using
 32-bit arithmetic.  This issue has been fixed by  typecasting
 to size_t; Thanks to John Baron.

 - 09/10/08 Version 2.0

 Piotr Luszczek changed to 64-bit RNG, modified files:
 -- [M] include/hpl_matgen.h
 -- [M] testing/matgen/HPL_ladd.c
 -- [M] testing/matgen/HPL_lmul.c
 -- [M] testing/matgen/HPL_rand.c
 -- [M] testing/ptest/HPL_pdinfo.c

 For a motivation for the change, see:
    Dongarra and Langou, ``The Problem with the Linpack
    Benchmark Matrix Generator'', LAWN 206, June 2008.

 -- [M] testing/ptest/HPL_pdtest.c  --

 Julien Langou changed the test for correctness from
      ||Ax-b||_oo / ( eps * ||A||_1  * N            )
      ||Ax-b||_oo / ( eps * ||A||_1  * ||x||_1      )
      ||Ax-b||_oo / ( eps * ||A||_oo * ||x||_oo * N )
 to the normwise backward error
      || r ||_oo / ( eps * ( || x ||_oo * || A ||_oo + || b ||_oo ) * N )
 See:
  Nicholas J. Higham, ``Accuracy and Stability of Numerical Algorithms'',
  Society for Industrial and Applied Mathematics, Philadelphia, PA, USA,
  Second Edition, pages = xxx+680, ISBN = 0-89871-521-0, 2002.

 Note that in our case || b ||_oo is almost for sure
 1/2, we compute it anyway.

 - 10/26/2012 Version 2.1

 Piotr Luszczek introduced exact time stamping for HPL_pdgesv():
 -- [M] dist/include/hpl_misc.h
 -- [M] dist/testing/ptest/HPL_pdtest.c

 Piotr Luszczek fixed out-of-bounds access in data spreading functions
 and exact time stamping for HPL_pdgesv():
 -- [M] dist/src/pgesv/HPL_spreadN.c
 -- [M] dist/src/pgesv/HPL_spreadT.c
 Thanks to Stephen Whalen from Cray.

 - 02/24/2016 Version 2.2

 Piotr Luszczek added continuous reporting of factorization progress
 submitted by Intel and make scripts that uses Intel software tools and
 libraries and their Apple's Mac OS X equivalents.


hpl-2.2/src/pauxil/HPL_pdlange.c
/* 
 * -- High Performance Computing Linpack Benchmark (HPL)                
 *    HPL - 2.2 - February 24, 2016                          
 *    Antoine P. Petitet                                                
 *    University of Tennessee, Knoxville                                
 *    Innovative Computing Laboratory                                 
 *    (C) Copyright 2000-2008 All Rights Reserved                       
 *                                                                      
 * -- Copyright notice and Licensing terms:                             
 *                                                                      
 * Redistribution  and  use in  source and binary forms, with or without
 * modification, are  permitted provided  that the following  conditions
 * are met:                                                             
 *                                                                      
 * 1. Redistributions  of  source  code  must retain the above copyright
 * notice, this list of conditions and the following disclaimer.        
 *                                                                      
 * 2. Redistributions in binary form must reproduce  the above copyright
 * notice, this list of conditions,  and the following disclaimer in the
 * documentation and/or other materials provided with the distribution. 
 *                                                                      
 * 3. All  advertising  materials  mentioning  features  or  use of this
 * software must display the following acknowledgement:                 
 * This  product  includes  software  developed  at  the  University  of
 * Tennessee, Knoxville, Innovative Computing Laboratory.             
 *                                                                      
 * 4. The name of the  University,  the name of the  Laboratory,  or the
 * names  of  its  contributors  may  not  be used to endorse or promote
 * products  derived   from   this  software  without  specific  written
 * permission.                                                          
 *                                                                      
 * -- Disclaimer:                                                       
 *                                                                      
 * THIS  SOFTWARE  IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,  INCLUDING,  BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE UNIVERSITY
 * OR  CONTRIBUTORS  BE  LIABLE FOR ANY  DIRECT,  INDIRECT,  INCIDENTAL,
 * SPECIAL,  EXEMPLARY,  OR  CONSEQUENTIAL DAMAGES  (INCLUDING,  BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA OR PROFITS; OR BUSINESS INTERRUPTION)  HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT,  STRICT LIABILITY,  OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 * ---------------------------------------------------------------------
 */ 
/*
 * Include files
 */
#include "hpl.h"

#ifdef STDC_HEADERS
double HPL_pdlange
(
   const HPL_T_grid *               GRID,
   const HPL_T_NORM                 NORM,
   const int                        M,
   const int                        N,
   const int                        NB,
   const double *                   A,
   const int                        LDA
)
#else
double HPL_pdlange
( GRID, NORM, M, N, NB, A, LDA )
   const HPL_T_grid *               GRID;
   const HPL_T_NORM                 NORM;
   const int                        M;
   const int                        N;
   const int                        NB;
   const double *                   A;
   const int                        LDA;
#endif
{
/* 
 * Purpose
 * =======
 *
 * HPL_pdlange returns  the value of the one norm,  or the infinity norm,
 * or the element of largest absolute value of a distributed matrix A:  
 *  
 *  
 *    max(abs(A(i,j))) when NORM = HPL_NORM_A,                          
 *    norm1(A),        when NORM = HPL_NORM_1,                          
 *    normI(A),        when NORM = HPL_NORM_I,                          
 *  
 * where norm1 denotes the one norm of a matrix (maximum column sum) and
 * normI denotes  the infinity norm of a matrix (maximum row sum).  Note
 * that max(abs(A(i,j))) is not a matrix norm.
 *
 * Arguments
 * =========
 *
 * GRID    (local input)                 const HPL_T_grid *
 *         On entry,  GRID  points  to the data structure containing the
 *         process grid information.
 *
 * NORM    (global input)                const HPL_T_NORM
 *         On entry,  NORM  specifies  the  value to be returned by this
 *         function as described above.
 *
 * M       (global input)                const int
 *         On entry,  M  specifies  the number  of rows of the matrix A.
 *         M must be at least zero.
 *
 * N       (global input)                const int
 *         On entry,  N specifies the number of columns of the matrix A.
 *         N must be at least zero.
 *
 * NB      (global input)                const int
 *         On entry,  NB specifies the blocking factor used to partition
 *         and distribute the matrix. NB must be larger than one.
 *
 * A       (local input)                 const double *
 *         On entry,  A  points to an array of dimension  (LDA,LocQ(N)),
 *         that contains the local pieces of the distributed matrix A.
 *
 * LDA     (local input)                 const int
 *         On entry, LDA specifies the leading dimension of the array A.
 *         LDA must be at least max(1,LocP(M)).
 *
 * ---------------------------------------------------------------------
 */ 
/*
 * .. Local Variables ..
 */
   double                     s, v0=HPL_rzero, * work = NULL;
   MPI_Comm                   Acomm, Ccomm, Rcomm;
   int                        ii, jj, mp, mycol, myrow, npcol, nprow,
                              nq;
/* ..
 * .. Executable Statements ..
 */
   (void) HPL_grid_info( GRID, &nprow, &npcol, &myrow, &mycol );
   Rcomm = GRID->row_comm; Ccomm = GRID->col_comm;
   Acomm = GRID->all_comm;

   Mnumroc( mp, M, NB, NB, myrow, 0, nprow );
   Mnumroc( nq, N, NB, NB, mycol, 0, npcol );

   if( Mmin( M, N ) == 0 ) { return( v0 ); }
   else if( NORM == HPL_NORM_A )
   {
/*
 * max( abs( A ) )
 */
      if( ( nq > 0 ) && ( mp > 0 ) )
      {
         for( jj = 0; jj < nq; jj++ )
         {
            for( ii = 0; ii < mp; ii++ )
            { v0 = Mmax( v0, Mabs( *A ) ); A++; }
            A += LDA - mp;
         }
      }
      (void) HPL_reduce( (void *)(&v0), 1, HPL_DOUBLE, HPL_max, 0,
                         Acomm );
   }
   else if( NORM == HPL_NORM_1 )
   {
/*
 * Find norm_1( A ).
 */
      if( nq > 0 )
      {
         work = (double*)malloc( (size_t)(nq) * sizeof( double ) );
         if( work == NULL )
         { HPL_pabort( __LINE__, "HPL_pdlange", "Memory allocation failed" ); }

         for( jj = 0; jj < nq; jj++ )
         {
            s = HPL_rzero;
            for( ii = 0; ii < mp; ii++ ) { s += Mabs( *A ); A++; }
            work[jj] = s; A += LDA - mp;
         }
/*
 * Find sum of global matrix columns, store on row 0 of process grid
 */
         (void) HPL_reduce( (void *)(work), nq, HPL_DOUBLE, HPL_sum,
                            0, Ccomm );
/*
 * Find maximum sum of columns for 1-norm
 */
         if( myrow == 0 )
         { v0 = work[HPL_idamax( nq, work, 1 )]; v0 = Mabs( v0 ); }
         if( work ) free( work );
      }
/*
 * Find max in row 0, store result in process (0,0)
 */
      if( myrow == 0 )
         (void) HPL_reduce( (void *)(&v0), 1, HPL_DOUBLE, HPL_max, 0,
                            Rcomm );
   }
   else if( NORM == HPL_NORM_I )
   {
/*
 * Find norm_inf( A )
 */
      if( mp > 0 )
      {
         work = (double*)malloc( (size_t)(mp) * sizeof( double ) );
         if( work == NULL )
         { HPL_pabort( __LINE__, "HPL_pdlange", "Memory allocation failed" ); }

         for( ii = 0; ii < mp; ii++ ) { work[ii] = HPL_rzero; }

         for( jj = 0; jj < nq; jj++ )
         {
            for( ii = 0; ii < mp; ii++ )
            { work[ii] += Mabs( *A ); A++; }
            A += LDA - mp;
         }
/*       
 * Find sum of global matrix rows, store on column 0 of process grid
 */      
         (void) HPL_reduce( (void *)(work), mp, HPL_DOUBLE, HPL_sum,
                            0, Rcomm );
/*       
 * Find maximum sum of rows for inf-norm
 */      
         if( mycol == 0 )
         { v0 = work[HPL_idamax( mp, work, 1 )]; v0 = Mabs( v0 ); }
         if( work ) free( work );
      }
/*
 * Find max in column 0, store result in process (0,0)
 */
      if( mycol == 0 )
         (void) HPL_reduce( (void *)(&v0), 1, HPL_DOUBLE, HPL_max,
                            0, Ccomm );
   }
/*
 * Broadcast answer to every process in the grid
 */
   (void) HPL_broadcast( (void *)(&v0), 1, HPL_DOUBLE, 0, Acomm );

   return( v0 );
/*
 * End of HPL_pdlange
 */
}
